### ISSUE TITLE
AsQueryable() on collection throws if applied after Where statement

### ISSUE THREAD

**Issue by lbcsy** [nhibernate/nhibernate-core#2471]
A query (IQueryable) worked in version 4.0.4 but has exception in 5.2.2:
```C#
 'query.ToList()' threw an exception of type 'System.NotSupportedException'
    Data: {System.Collections.ListDictionaryInternal}
    HResult: -2146233067
    HelpLink: null
    InnerException: null
    Message: "query ( query ( select_from ( from ( range ( . _0 RolePrivileges ) a ) ) ( select a ) ) ( where ( and ( == ( . a AccessTypeName ) ( : p12 ) ) ( or ( == ( : p13 ) true ) ( == ( . ( . param002 Role ) Id ) ( : p14 ) ) ) ) ) )"
    Source: "NHibernate"
    StackTrace: "   at NHibernate.Hql.Ast.ANTLR.PolymorphicQuerySourceDetector.GetClassName(IASTNode querySource)\r\n   at NHibernate.Hql.Ast.ANTLR.PolymorphicQuerySourceDetector.Process(IASTNode tree)\r\n   at NHibernate.Hql.Ast.ANTLR.AstPolymorphicProcessor.Process()\r\n   at NHibernate.Hql.Ast.ANTLR.ASTQueryTranslatorFactory.CreateQueryTranslators(IASTNode ast, String queryIdentifier, String collectionRole, Boolean shallow, IDictionary`2 filters, ISessionFactoryImplementor factory)\r\n   at NHibernate.Engine.Query.QueryPlanCache.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow, IDictionary`2 enabledFilters)\r\n   at NHibernate.Impl.AbstractSessionImpl.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow)\r\n   at NHibernate.Impl.AbstractSessionImpl.CreateQuery(IQueryExpression queryExpression)\r\n   at NHibernate.Linq.DefaultQueryProvider.PrepareQuery(Expression expression, IQuery& query)\r\n   at NHibernate.Linq.DefaultQueryProvider.Execute(Expression expression)\r\n   at NHib
ernate.Linq.DefaultQueryProvider.Execute[TResult](Expression expression)\r\n   at Remotion.Linq.QueryableBase`1.GetEnumerator()\r\n   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)\r\n   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)"
```

Can be reproduced by adding the following test to `WhereSubqueryTests`
```C#
[Test(Description = "GH-2471")]
public void TimeSheetsWithStringContainsSubQueryWithAsQueryableAfterWhere()
{
	var query = (from timesheet in db.Timesheets
				where timesheet.Entries.Where(e => e.Comments != null).AsQueryable().Any(e => e.Comments.Contains("testing"))
				select timesheet).ToList();

	Assert.That(query.Count, Is.EqualTo(2));
}
```

**Comment by fredericDelaporte** [comment]
Hello, you have reached the issue tracker, meant for bug reports or feature requests.

Please post on the [nhusers group](https://groups.google.com/forum/#!forum/nhusers) for support requests (or see [Groups](https://nhibernate.info/groups.html)). (And better provide more information while doing so, I doubt someone will answer you without at least the LINQ query provided.)

**Comment by bahusoid** [comment]
Actually it looks like some real issue as there is no known limitations in 5.x compared to 4.x. Only it needs proper reporting (class mappings and failed query example)  Also please make sure it's still reproducible on latest 5.2 release (currently 5.2.7) and on recent 5.3 release.

**Comment by lbcsy** [comment]
It fails in both 5.2 and 5.3.
I can now identify the error spot, but don't know why and how to overcome it:

The query seems failed with IEnumerable.AsQueryable:
```
var predicateReadPrivilege = PredicateBuilder.Create<Resource>(x =>
                x.RType.RolePrivileges.Where(a => a.AccessTypeName == accessLevel).AsQueryable().Where(_ => true).Any());// _predicatePrivilegeOnResourceType).Any());
```
As you can see, even the criteria is a redundant one, it fails, but if I remove the AsQueryable part, it does not throw the exception.

**Comment by bahusoid** [comment]
It should work if you apply `AsQueryable` right after mapped collection `x.RType.RolePrivileges.AsQueryable()...`

Yeah if such query works in 4.x it looks like a bug. Can be reproduced adding the following test to `WhereSubqueryTests`
```C#
[Test(Description = "GH-2471")]
public void TimeSheetsWithStringContainsSubQueryWithAsQueryableAfterWhere()
{
	var query = (from timesheet in db.Timesheets
				where timesheet.Entries.Where(e => e.Comments != null).AsQueryable().Any(e => e.Comments.Contains("testing"))
				select timesheet).ToList();

	Assert.That(query.Count, Is.EqualTo(2));
}
```

**Comment by fredericDelaporte** [comment]
> Yeah if such query works in 4.x it looks like a bug.

Actual support for `AsQueryable` was added in v5.0 with #375, causing the query to actually occur on db-side rather than in-memory, without even loading the collection when lazy and not yet loaded. With 4.x, this was yielding instead a linq-to-object queryable.

I do not really see the point of calling `AsQueryable` after having already call some enumerable Linq functions: it looks to me as a coding error.

For me, this case looks more as maybe a forgotten possible breaking change in v5.0.

**Comment by bahusoid** [comment]
> Actual support for AsQueryable was added in v5.0 with #375, 

No. It's different issue, Your issue is about collection filters. And this issue about `AsQueryable` call inside lambda expressions (and it doesn't involve in-memory calculations). It's been added in v3.3 in https://nhibernate.jira.com/browse/NH-2998 

Yeah I also don't think it's a minor issue but still if such query works in 4.x it's a bug.

**Comment by fredericDelaporte** [comment]
If this is a regression of 5.0, shouldn't we backport its fix down to 5.0.x?

**Comment by hazzik** [comment]
Yes, we should. But do we want to do it?
